client/scala/src/main/scala/ai/verta/blobs/dataset/S3.scala

```diff
@@ -0,0 +1,168 @@
+
+    queryAttempt match {
+      case Failure(e) => Failure(e)
+      case Success(list) => Success(new S3(
+        list.map(metadata => metadata.path -> metadata),
+        list.filter(_.versionId.isDefined).map(metadata => metadata.path -> metadata.versionId.get)
+      ))
+    }
+  }
+
+  /** Factory method to convert a versioning blob instance
+   *  @param s3VersioningBlob the versioning blob to convert
+   */
+  def apply(s3VersioningBlob: VersioningS3DatasetBlob) {
+    var s3Blob = new S3(List(), List())
+    val componentList = s3VersioningBlob.components.get
+    val metadataList = componentList.map(
+      comp => comp.path.get.path.get -> s3Blob.toMetadata(comp.path.get)
+    )
+    val versionList = componentList.filter(_.s3_version_id.isDefined).map(
+      comp => comp.path.get.path.get -> comp.s3_version_id.get
+    )
+
+    s3Blob.contents = HashMap(metadataList: _*)
+    s3Blob.versionMap = HashMap(versionList: _*)
+
+    s3Blob
+  }
+
+  /** Convert a S3 instance to a VersioningBlob
+   *  @param blob S3 instance
+   *  @return equivalent VersioningBlob instance
+   */
+  def toVersioningBlob(blob: S3) = {
+    val s3Components = blob.components.map(comp => VersioningS3DatasetComponentBlob(
+      Some(comp),
+      blob.versionMap.get(comp.path.get)
+    ))
+
+    VersioningBlob(
+      dataset = Some(VersioningDatasetBlob(
+        s3 = Some(VersioningS3DatasetBlob(Some(s3Components)))
+      ))
+    )
+  }
+
+  /** Helper function to query metadata of S3 Location
+   */
+  private def getS3LocMetadata(loc: S3Location) = {
+    if (loc.key.isEmpty || loc.key.get.endsWith("/")) {
+      // no key (bucket), or is a folder
+      val versionListing =
+        if (loc.key.isEmpty) s3.listVersions(loc.bucketName, null)
+        else s3.listVersions(loc.bucketName, loc.key.get)
+
+      handleVersionListing(versionListing, List())
+    }
+    else {
+      val request =
+        if (loc.versionID.isDefined) new GetObjectMetadataRequest(loc.bucketName, loc.key.get, loc.versionID.get)
+        else new GetObjectMetadataRequest(loc.bucketName, loc.key.get)
+
+      getObjectMetadata(s3.getObjectMetadata(request), loc.bucketName, loc.key.get).map(List(_))
+    }
+  }
+
+  /** Helper function to deal with VersionListing
+   */
+  @tailrec private def handleVersionListing(
+    versionListing: VersionListing,
+    acc: List[FileMetadata]
+  ): Try[List[FileMetadata]] = {
+    val batchAttempt = Try(
+      versionListing.getVersionSummaries().asScala.toList
+                    .filter((version: S3VersionSummary) => !version.getKey().endsWith("/")) // not a folder
+                    .filter(_.isLatest())
+                    .map(getVersionMetadata) // List[Try]
+                    .map(_.get)
+      )
+
+    batchAttempt match {
+      case Failure(e) => Failure(e)
+      case Success(batch) =>
+        if (versionListing.isTruncated())
+          handleVersionListing(s3.listNextBatchOfVersions(versionListing), batch ::: acc)
```

Comment:
Maintain order. It might matter.

Reply:
Hmm Ok. I'll reverse batch here, which makes the final list in reverse order:
(3, 4).reverse :::  (1, 2).reverse = (4, 3, 2, 1)
Finally, I'll reverse the result list once more. This will give the correct order.
Note that the order of concatenation is for efficiency reason. I believe that it's better to prepending the new stuff rather than appending them?

Reply:
Should be fixed, but do tell me if there's a better way.

Reply:
Why do you have to reverse? Why doesn't `acc ::: batch` work?

Reply:
Resolved in chat. Added clarifying comments.